### ISSUE TITLE
0‘“  Resulting in syntax highlighting confusion

### ISSUE THREAD

**Issue by JIM-GLITCH** [jeff-hykin/better-prolog-syntax#1]
I like your plugin. There is a small flaw. `0'"` results in syntax highlighting confusion. In prolog, `0'"` means 34, which is the ascii value of the character `"`. But the color from this `"` to the next `"` is the color of the string.

**Comment by jeff-hykin** [comment]
Thanks for posting about the issue!

And well the highlighter isn't the only one who is confused 😅. I don't understand how Prolog is parsing that. Could you provide some more examples so I could see a pattern? (Or a specification describing the feature)

**Comment by JIM-GLITCH** [comment]
I can't find a specification describing the feature. Forget where it is😅. `0'` is followed by a normal character or an escape character. SWI-Prolog  reference manual 2.16.1.3 Character Escape Syntax :
[https://www.swi-prolog.org/pldoc/man?section=syntax](url)
examples :
```prolog
% normal character
?- display(0'1).
49
?- display(0'a).
97
?- display(0'.).
46
?- display(0'#).
35
?- display(0'?).
63
?- display(0'').
39
?- display(0'").
34
% escape character.
?- display(0'\123).
83
?- display(0'\x123).
291
?- display(0'\u1234).
4660
?- display(0'\U00001234).
4660
```
I know how to parse it with prolog. This is my file.
```prolog
:- use_module(library(dcg/basics)).


zero_apostrophe(I)-->
    "0'",
    char_num(I).
char_num(Val)-->
    [C],
    (
    {C==0'\\}
    ->  look_ahead(X),
        (
            {X=<0'7,code_type(X,digit)} 
        ->  octonary_integer_exp_ass(Val)
        ;   {X==0'x}
        ->  [_],hexadecimal_integer_exp_ass(Val)
        ;   {X==0'u}
        ->  [_],hexadecimal_integer_exp_ass(Val,4) % 4-digit hexadecimal number
        ;   {X==0'U}
        ->  [_],hexadecimal_integer_exp_ass(Val,8), % 8-digit hexadecimal number
            {Val =< 65535}
        ;   [_],{esc_ascii(X,Val)}
        )
    ; {Val=C}
    ).

    
esc_ascii(0'a,7).
esc_ascii(0'b,8).
esc_ascii(0'c,99).
esc_ascii(0'e,27).
esc_ascii(0'f,12).
esc_ascii(0'n,10).
esc_ascii(0'r,13).
esc_ascii(0's,32).
esc_ascii(0't,9).
esc_ascii(0'v,11).
esc_ascii(0'',39).
esc_ascii(0'" ,34). 
esc_ascii(0'\\,92).

hexadecimal_integer_exp_ass(Val)-->
    xdigit(D0),
    !,
	xdigits(D),
	{ mkval([D0|D], 16, Val)
	}.
hexadecimal_integer_exp_ass(Val,Length)-->
    xdigit(D0),
    !,
	xdigits(D),
	{ length([D0|D],Length),mkval([D0|D], 16, Val)
	}.

octonary_integer_exp_ass(Val)-->
    octonary_digit(D0),
    !,
    octonary_digits(D),
    { mkval([D0|D], 8, Val)
    }.

look_ahead(X), [X] --> [X]. 

% Base conversion
mkval([W0|Weights], Base, Val) :-
	mkval(Weights, Base, W0, Val).

mkval([], _, W, W).
mkval([H|T], Base, W0, W) :-
	W1 is W0*Base+H,
	mkval(T, Base, W1, W).

octonary_digit(D)-->
    [C],
    {C=<0'7,
    code_type(C,digit(D))
    }.

octonary_digits([D0|D])-->
    octonary_digit(D0),
    !,
    octonary_digits(D).
octonary_digits([])-->
    [].
```
It works fine.
```prolog
% \ In ``  needs to be escaped
?- zero_apostrophe(A,`0'1`,R).             
A = 49,
R = [].
?- zero_apostrophe(A,`0'a`,R). 
A = 97,
R = [].
?- zero_apostrophe(A,`0'\\\\`,R). 
A = 92,
R = [].
?- zero_apostrophe(A,`0'\\u1234`,R). 
A = 4660,
R = [].
?- zero_apostrophe(A,`0'\\U00001234`,R). 
A = 4660,
R = [].
?- zero_apostrophe(A,`0'\\"`,R).         
A = 34,
R = [].
?- zero_apostrophe(A,`0'\\'`,R). 
A = 39,
R = [].
```

**Comment by jeff-hykin** [comment]
Thanks that helps a lot. 

I added the patterns [here](https://github.com/jeff-hykin/better-prolog-syntax/blob/884681aa14ad2e6197e21416287a9be96e317d01/main/patterns/character.rb#L16) (look at the # comments and the `should_fully_match` and `should_not_partial_match ` examples)

New version is published `0.3.1`

Let me know if something was missed
